### ISSUE TITLE
Potential reportPossiblyUnboundVariable refinement

### ISSUE THREAD

**Issue by OfekShilon** [microsoft/pyright#11101]
```python
def f(data: list[int]) -> bool:
    if not data:
        return False

    for num in data:
        ret = num > 5

    return ret   # "ret" is possibly unbound  (reportPossiblyUnboundVariable)
```

This pyright error is bogus, perhaps it's solvable?

**Comment by erictraut** [comment]
The Python type system doesn't provide enough information to statically determine the length of an `Iterable`. Pyright has separate rules for `reportUnboundVariable` and `reportPossiblyUnboundVariable`. The latter will (necessarily) have some false positives. If you find it too noisy, you can disable it. I generally recommend initializing local variables outside of a loop if you plan to refer to them after the loop.
